main.go

```diff
@@ -118,93 +143,113 @@ type writer interface {
 	Name() string
 }
 
+type noOpWriter struct{}
+
+func (no *noOpWriter) Write(samples model.Samples) error {
+	util.LogDebug("msg", "Noop writer", "num_samples", len(samples))
+	return nil
+}
+
+func (no *noOpWriter) Name() string {
+	return "noopWriter"
+}
+
 type reader interface {
 	Read(req *prompb.ReadRequest) (*prompb.ReadResponse, error)
 	Name() string
 	HealthCheck() error
 }
 
-func buildClients(logger log.Logger, cfg *config) ([]writer, []reader) {
-	var writers []writer
-	var readers []reader
+func buildClients(cfg *config) (writer, reader) {
+	var writer writer
+	var reader reader
 
-	pgClient := pgprometheus.NewClient(logger, &cfg.pgPrometheusConfig)
-	writers = append(writers, pgClient)
-	readers = append(readers, pgClient)
-
-	return writers, readers
+	pgClient := pgprometheus.NewClient(&cfg.pgPrometheusConfig)
+	reader = pgClient
+	if cfg.readOnly {
+		writer = &noOpWriter{}
+	} else {
+		writer = pgClient
+	}
+	return writer, reader
 }
 
-func serve(logger log.Logger, addr string, writers []writer, readers []reader) error {
-	http.HandleFunc("/write", func(w http.ResponseWriter, r *http.Request) {
+func write(writer writer) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		compressed, err := ioutil.ReadAll(r.Body)
 		if err != nil {
-			level.Error(logger).Log("msg", "Read error", "err", err.Error())
+			util.LogError("msg", "Read error", "err", err.Error())
 			http.Error(w, err.Error(), http.StatusInternalServerError)
 			return
 		}
 
 		reqBuf, err := snappy.Decode(nil, compressed)
 		if err != nil {
-			level.Error(logger).Log("msg", "Decode error", "err", err.Error())
+			util.LogError("msg", "Decode error", "err", err.Error())
 			http.Error(w, err.Error(), http.StatusBadRequest)
 			return
 		}
 
 		var req prompb.WriteRequest
 		if err := proto.Unmarshal(reqBuf, &req); err != nil {
-			level.Error(logger).Log("msg", "Unmarshal error", "err", err.Error())
+			util.LogError("msg", "Unmarshal error", "err", err.Error())
 			http.Error(w, err.Error(), http.StatusBadRequest)
 			return
 		}
 
 		samples := protoToSamples(&req)
 		receivedSamples.Add(float64(len(samples)))
 
-		var wg sync.WaitGroup
-		for _, w := range writers {
-			wg.Add(1)
-			go func(rw writer) {
-				sendSamples(logger, rw, samples)
-				wg.Done()
-			}(w)
+		sendSamples(writer, samples)
+		counter, err := sentSamples.GetMetricWithLabelValues(writer.Name())
+		if err != nil {
+			util.LogWarn("msg", "Couldn't get a counter", "labelValue", writer.Name(), "err", err)
+		}
+		writeThroughtput.SetCurrent(getCounterValue(counter))
+
+		select {
+		case d := <-writeThroughtput.Values:
+			util.LogInfo("msg", "Samples write throughput", "samples/sec", d)
```

Comment:
I think this should be logged at DEBUG level, because otherwise this will ultimately fill the log and there is no way to run at INFO level without per-request log output.

Reply:
Yes ,I thought about it, but since `writeThroughout` calculation runs every second I feel it's not going to add a lot of noise.